module/zfs/arc.c

```diff
@@ -8687,6 +9020,785 @@ l2arc_stop(void)
+		vdev_space_update(dev->l2ad_vdev, ipsize, 0, 0);
+		l2arc_hdr_restore(&lb->lb_entries[i], dev, load_guid);
+	}
+
+	/*
+	 * Record rebuild stats:
+	 *	size		In-memory size of restored buffer data in ARC
+	 *	psize		Physical size of restored buffers in the L2ARC
+	 *	bufs		# of ARC buffer headers restored
+	 *	log_blks	# of L2ARC log entries processed during restore
+	 */
+	ARCSTAT_INCR(arcstat_l2_rebuild_size, size);
+	ARCSTAT_INCR(arcstat_l2_rebuild_psize, psize);
+	ARCSTAT_INCR(arcstat_l2_rebuild_bufs, L2ARC_LOG_BLK_ENTRIES);
+	ARCSTAT_BUMP(arcstat_l2_rebuild_log_blks);
+	ARCSTAT_F_AVG(arcstat_l2_log_blk_avg_size, lb_psize);
+	ARCSTAT_F_AVG(arcstat_l2_data_to_meta_ratio, psize / lb_psize);
+}
+
+/*
+ * Restores a single ARC buf hdr from a log block. The ARC buffer is put
+ * into a state indicating that it has been evicted to L2ARC.
+ */
+static void
+l2arc_hdr_restore(const l2arc_log_ent_phys_t *le, l2arc_dev_t *dev,
+    uint64_t load_guid)
+{
+	arc_buf_hdr_t		*hdr, *exists;
+	kmutex_t		*hash_lock;
+	arc_buf_contents_t	type = BLKPROP_GET_TYPE((le)->le_prop);
+	uint64_t		asize;
+
+	/*
+	 * Do all the allocation before grabbing any locks, this lets us
+	 * sleep if memory is full and we don't have to deal with failed
+	 * allocations.
+	 */
+	hdr = arc_buf_alloc_l2only(load_guid, BLKPROP_GET_LSIZE((le)->le_prop),
+	    type, dev, le->le_dva, le->le_daddr,
+	    BLKPROP_GET_PSIZE((le)->le_prop), le->le_birth,
+	    BLKPROP_GET_COMPRESS((le)->le_prop), le->le_protected);
+	asize = arc_hdr_size(hdr);
+
+	ARCSTAT_INCR(arcstat_l2_lsize, HDR_GET_LSIZE(hdr));
+	ARCSTAT_INCR(arcstat_l2_psize, asize);
+
+	mutex_enter(&dev->l2ad_mtx);
+	/*
+	 * We connect the l2hdr to the hdr only after the hdr is in the hash
+	 * table, otherwise the rest of the arc hdr manipulation machinery
+	 * might get confused.
+	 */
+	list_insert_tail(&dev->l2ad_buflist, hdr);
+	(void) zfs_refcount_add_many(&dev->l2ad_alloc, asize, hdr);
+	mutex_exit(&dev->l2ad_mtx);
+
+	exists = buf_hash_insert(hdr, &hash_lock);
+	if (exists) {
+		/* Buffer was already cached, no need to restore it. */
+		arc_hdr_destroy(hdr);
+		mutex_exit(hash_lock);
+		ARCSTAT_BUMP(arcstat_l2_rebuild_bufs_precached);
+		return;
+	}
+
+	mutex_exit(hash_lock);
+}
+
+/*
+ * Starts an asynchronous read IO to read a log block. This is used in log
+ * block reconstruction to start reading the next block before we are done
+ * decoding and reconstructing the current block, to keep the l2arc device
+ * nice and hot with read IO to process.
+ * The returned zio will contain a newly allocated memory buffers for the IO
+ * data which should then be freed by the caller once the zio is no longer
+ * needed (i.e. due to it having completed). If you wish to abort this
+ * zio, you should do so using l2arc_log_blk_prefetch_abort, which takes
+ * care of disposing of the allocated buffers correctly.
+ */
+static zio_t *
+l2arc_log_blk_prefetch(vdev_t *vd, const l2arc_log_blkptr_t *lbp,
+    uint8_t *lb_buf)
```

Comment:
To "prefetch" means to read some data into a cache, with the intent that a later cache access will get a cache hit.  It doesn't look like that's what's happening here.  This is simply issuing an asynchronous read which will fill the provided buffer.  The caller waits for returned zio to complete, at which point the provided buffer is filled with the requested data.

Reply:
I think it was named as such according to the "prefetch" comment in line 7592 (see above). We can call it just fetch for now.